modules/files.nix

```diff
@@ -0,0 +1,103 @@
+{ lib, pkgs, ... }:
+let
+  fileType = lib.types.submodule (
+    {
+      name,
+      config,
+      options,
+      ...
+    }:
+    {
+      options = {
+        enable = lib.mkOption {
+          type = lib.types.bool;
+          default = true;
+          description = ''
+            Whether this file should be generated.
+            This option allows specific files to be disabled.
+          '';
+        };
+
+        target = lib.mkOption {
+          type = lib.types.str;
+          defaultText = lib.literalMD "the attribute name";
+          description = ''
+            Name of symlink, relative to nvim config.
+          '';
+        };
+
+        text = lib.mkOption {
+          type = with lib.types; nullOr lines;
+          default = null;
+          description = "Text of the file.";
+        };
+
+        source = lib.mkOption {
+          type = lib.types.path;
+          description = "Path of the source file.";
+        };
+      };
+
+      config =
+        let
+          derivationName = "nvim-" + lib.replaceStrings [ "/" ] [ "-" ] name;
+        in
+        {
+          target = lib.mkDefault name;
+          source = lib.mkIf (config.text != null) (
+            # mkDerivedConfig uses the option's priority, and calls our function with the option's value.
+            # This means our `source` definition has the same priority as `text`.
+            lib.mkDerivedConfig options.text (pkgs.writeText derivationName)
+          );
+        };
+    }
+  );
+
+  # TODO: Added 2024-07-07, remove after 24.11
+  # Before we had a fileType, we used types.str.
+  # We coerce to the new type with a warning.
+  coercedFileType =
+    let
+      oldType = lib.types.str;
+      newType = fileType;
+      fn = text: { inherit text; };
+    in
+    (lib.types.coercedTo oldType fn newType)
+    // {
+      # Don't show oldType in the description
+      inherit (newType) description;
+
+      # Override merge to print a warning when oldType is used
+      merge =
+        opt: defs:
+        let
+          coerceVal =
+            val:
+            if oldType.check val then
+              lib.warn ''
+                Passing a ${oldType.description} for ${lib.showOption opt} is deprecated, use ${newType.description} instead. Definitions: ${lib.options.showDefs defs}
+              '' (fn val)
```

Comment:
I don't see this warning when running:
```
nix build .#checks.x86_64-linux.tests.passthru.entries.modules-extra-files-query
```
So I suspect something isn't working...

Reply:
Resolved: overriding attributes on the type returned by `types.coercedTo` wasn't working, but copying its source and making the appropriate changes does work.